File: src/python/engine/anki.py
import sqlite3
from zipfile import ZipFile
from tempfile import mkdtemp
import hashlib
from typing import Callable, Any, Union
import os
import json
from datetime import datetime
from pathlib import Path
import re
import pystache

from .db import Db


class Anki:
    def __init__(self, file_path: str, true_filename: str, cb: Callable[[Union[dict, None]], Any]):
        self.filename = true_filename
        self.dir = mkdtemp()
        self.cb = cb

        with ZipFile(file_path) as zf:
            fs = zf.namelist()

            for i, f in enumerate(fs):
                self.cb({
                    "text": f"Unzipping {f}",
                    "current": i,
                    "max": len(fs)
                })

                zf.extract(f, self.dir)

        self.cb({
            "text": "Preparing Anki resources"
        })

        self.conn = sqlite3.connect(os.path.join(self.dir, "collection.anki2"), check_same_thread=False)
        self.conn.row_factory = sqlite3.Row

        self.conn.executescript("""
        CREATE TABLE decks (
            id      INTEGER PRIMARY KEY,
            name    VARCHAR UNIQUE NOT NULL
        );

        CREATE TABLE models (
            id      INTEGER PRIMARY KEY,
            name    VARCHAR NOT NULL,
            flds    VARCHAR NOT NULL,
            css     VARCHAR
        );

        CREATE TABLE templates (
            id      INTEGER PRIMARY KEY AUTOINCREMENT,
            mid     INTEGER REFERENCES model(id),
            name    VARCHAR NOT NULL,
            qfmt    VARCHAR NOT NULL,
            afmt    VARCHAR
        );
        """)

        d_text, m_text = self.conn.execute("""
        SELECT decks, models FROM col
        """).fetchone()

        decks = json.loads(d_text)
        models = json.loads(m_text)

        assert isinstance(decks, dict)

        for d in decks.values():
            self.conn.execute("""
            INSERT INTO decks (id, name)
            VALUES (?, ?)
            """, (int(d["id"]), d["name"]))

        assert isinstance(models, dict)

        for m in models.values():
            self.conn.execute("""
            INSERT INTO models (id, name, flds, css)
            VALUES (?, ?, ?, ?)
            """, (
                int(m["id"]),
                m["name"],
                "\x1f".join(f["name"] for f in m["flds"]),
                m["css"]
            ))

            for t in m["tmpls"]:
                self.conn.execute("""
                INSERT INTO templates (mid, name, qfmt, afmt)
                VALUES (?, ?, ?, ?)
                """, (int(m["id"]), t["name"], t["qfmt"], t["afmt"]))

    def close(self):
        self.conn.close()
        os.unlink(self.dir)

    def export(self, db: Db) -> None:
        self.cb({
            "text": "Writing to database"
        })

        source_id = db.conn.execute("""
        INSERT INTO source (name, h, created)
        VALUES (?, ?, ?)
        """, (
            self.filename,
            hashlib.md5(Path(self.filename).read_bytes()).hexdigest(),
            str(datetime.now())
        ))

        media_name_to_id = dict()
        media_json = json.load(os.path.join(self.dir, "media"))

        assert isinstance(media_json, dict)
        for i, k in enumerate(media_json.keys()):
            data = Path(self.dir).joinpath(k).read_bytes()
            h = hashlib.md5(data).hexdigest()

            self.cb({
                "text": "Uploading media",
                "current": i,
                "max": len(media_json)
            })

            db.conn.execute("""
            INSERT INTO media (sourceId, name, data, h)
            VALUES (?, ?, ?, ?)
            ON CONFLICT DO NOTHING
            """, (source_id, media_json[k], data, h))

            media_name_to_id[media_json[k]] = db.conn.execute("""
            SELECT id FROM media
            WHERE h = ?
            """, (h,)).fetchone()[0]

        ts = self.conn.execute("""
        SELECT t.name AS tname, m.name AS mname, qfmt, afmt, css
        FROM templates AS t
        INNER JOIN models AS m ON m.id = t.mid
        """).fetchall()

        for i, t in enumerate(ts):
            self.cb({
                "text": "Uploading templates",
                "current": i,
                "max": len(ts)
            })

            db.conn.execute("""
            INSERT INTO template (name, model, front, back, css, sourceId)
            VALUES (?, ?, ?, ?, ?, ?)
            """, (
                t["tname"],
                t["mname"],
                self._convert_link(t["qfmt"], media_name_to_id),
                self._convert_link(t["afmt"], media_name_to_id),
                self._convert_link(t["css"], media_name_to_id),
                source_id
            ))

        front_set = set()
        ns = self.conn.execute("""
        SELECT
            n.flds AS "values",
            m.flds AS keys,
            t.name AS tname,
            m.name AS mname,
            d.name AS deck,
            qfmt,
            tags
        FROM cards AS c
        INNER JOIN decks AS d ON d.id = did
        INNER JOIN notes AS n ON n.id = nid
        INNER JOIN models AS m ON m.id = n.mid
        INNER JOIN templates AS t ON t.mid = n.mid
        """).fetchall()

        for i, n in enumerate(ns):
            self.cb({
                "text": "Uploading notes",
                "current": i,
                "max": len(ns)
            })

            vs = n["values"].split("\x1f")
            ks = n["keys"].split("\x1f")
            data = dict(zip(ks, vs))

            front = pystache.render(n["qfmt"], data)
            if front == pystache.render(n["qfmt"], dict()):
                continue

            front = f"md5\n{hashlib.md5(front.encode()).hexdigest()}"
            if front in front_set:
                continue

            front_set.add(front)

            db.insert_many([{
                "deck": n["deck"].replace("::", "/"),
                "model": n["mname"],
                "template": n["tname"],
                "entry": vs[0],
                "data": data,
                "front": front,
                "tag": n["tag"].split(" "),
                "sourceId": source_id
            }])

        self.cb(None)

    @staticmethod
    def _convert_link(s: str, media_name_to_id: dict) -> str:
        return re.sub(
            "(?:(?:href|src)=\")([^\"]+)(?:\")",
            lambda m: f"/media/{media_name_to_id[m[1]]}",
            s
        )

File: src/node/engine/anki.ts
import sql from "sql.js";
import fs from "fs";
import AdmZip from "adm-zip";
import path from "path";
import Db, { ITemplate, IInsertEntry, IMedia } from "./db";
import crypto from "crypto";
import { simpleMustacheRender } from "../util";

export default class Anki {
    private db: sql.Database;
    private mediaNameToId: any = {};
    private filename: string;
    private filepath: string;
    private dir: string;
    private callback: (res: any) => any;

    constructor(filepath: string, filename: string, callback: (res: any) => any) {
        this.filename = filename;
        this.filepath = filepath;
        this.dir = path.dirname(filepath);
        this.callback = callback;

        const zip = new AdmZip(filepath);
        const zipCount = zip.getEntries().length;

        this.callback({
            text: `Unzipping Apkg. File count: ${zipCount}`,
            max: 0
        });

        zip.extractAllTo(this.dir);

        this.callback({
            text: "Preparing Anki resources.",
            max: 0
        });

        this.db = new sql.Database(fs.readFileSync(path.join(this.dir, "collection.anki2")));

        const { decks, models } = (() => {
            const st = this.db.prepare("SELECT decks, models FROM col");
            st.step();
            return st.getAsObject();
        })();

        this.db.run(`
        CREATE TABLE decks (
            id      INTEGER NOT NULL PRIMARY KEY,
            name    VARCHAR NOT NULL
        )`);

        const stmt = this.db.prepare("INSERT INTO decks (id, name) VALUES (?, ?)");

        Object.values(JSON.parse(decks as string)).forEach((deck: any) => {
            stmt.run([deck.id, deck.name]);
        });

        this.db.run(`
        CREATE TABLE models (
            id      INTEGER NOT NULL PRIMARY KEY,
            name    VARCHAR NOT NULL,
            flds    VARCHAR NOT NULL,
            css     VARCHAR
        )`);

        this.db.run(`
        CREATE TABLE templates (
            id      INTEGER PRIMARY KEY AUTOINCREMENT,
            mid     INTEGER REFERENCES model(id),
            name    VARCHAR NOT NULL,
            qfmt    VARCHAR NOT NULL,
            afmt    VARCHAR
        )`);

        const modelInsertStmt = this.db.prepare("INSERT INTO models (id, name, flds, css) VALUES (?, ?, ?, ?)");
        const templateInsertStmt = this.db.prepare("INSERT INTO templates (mid, name, qfmt, afmt) VALUES (?, ?, ?, ?)");

        Object.values(JSON.parse(models as string)).forEach((model: any) => {
            modelInsertStmt.run([model.id, model.name, model.flds.map((f: any) => f.name).join("\x1f"), model.css]);

            model.tmpls.forEach((t: any) => {
                templateInsertStmt.run([model.id, t.name, t.qfmt, t.afmt]);
            });
        });
    }

    public export(dst: Db) {
        this.callback({
            text: "Writing to database",
            max: 0
        });

        let sourceId: number;
        try {
            sourceId = dst.source.insertOne({
                name: this.filename,
                h: md5hasher(fs.readFileSync(this.filepath)),
                created: new Date()
            }).$loki;
        } catch (e) {
            this.callback({
                error: `Duplicated resource: ${this.filename}`
            });
            return;
        }

        this.mediaNameToId = {} as any;

        const mediaJson = JSON.parse(fs.readFileSync(path.join(this.dir, "media"), "utf8"));

        Object.keys(mediaJson).forEach((k, i) => {
            const data = fs.readFileSync(path.join(this.dir, k));
            const h = md5hasher(data);
            const media = {
                sourceId,
                name: mediaJson[k],
                data,
                h
            } as IMedia;

            const total = Object.keys(mediaJson).length;
            this.callback({
                text: "Uploading media",
                current: i,
                max: total
            });

            let mediaId;
            try {
                mediaId = dst.media.insert(media).$loki;
            } catch (e) {
                mediaId = dst.media.findOne({h}).$loki;
            }

            this.mediaNameToId[media.name] = mediaId;
        });

        const templates = [] as ITemplate[];

        (() => {
            const stmt = this.db.prepare(`
            SELECT t.name AS tname, m.name AS mname, qfmt, afmt, css
            FROM templates AS t
            INNER JOIN models AS m ON m.id = t.mid`);

            while (stmt.step()) {
                const { tname, mname, qfmt, afmt, css } = stmt.getAsObject();
                templates.push({
                    name: tname as string,
                    model: mname as string,
                    front: this.convertLink(qfmt as string),
                    back: this.convertLink(afmt as string),
                    css: this.convertLink(css as string),
                    sourceId
                });
            }
        })();

        (() => {
            const batch = 1000;
            const total = templates.length;
            let subList = templates.splice(0, batch);
            let from = 0;

            while (subList.length > 0) {
                this.callback({
                    text: "Uploading templates",
                    current: from,
                    max: total
                });

                dst.template.insert(subList);
                subList = templates.splice(0, batch);
                from += batch;
            }
        })();

        const count = (() => {
            let i = 0;
            const stmt = this.db.prepare(`
            SELECT
                COUNT(*)
            FROM cards AS c
            INNER JOIN decks AS d ON d.id = did
            INNER JOIN notes AS n ON n.id = nid
            INNER JOIN models AS m ON m.id = n.mid
            INNER JOIN templates AS t ON t.mid = n.mid`);
            stmt.step();
            return stmt.get()[0] as number;
        })();

        const entries = [] as IInsertEntry[];
        const frontSet = new Set();

        (() => {
            const stmt = this.db.prepare(`
            SELECT
                n.flds AS "values",
                m.flds AS keys,
                t.name AS tname,
                m.name AS mname,
                d.name AS deck,
                qfmt,
                tags
            FROM cards AS c
            INNER JOIN decks AS d ON d.id = did
            INNER JOIN notes AS n ON n.id = nid
            INNER JOIN models AS m ON m.id = n.mid
            INNER JOIN templates AS t ON t.mid = n.mid`);

            let current = 0;
            while (stmt.step()) {
                if (!(current % 1000)) {
                    this.callback({
                        text: "Reading notes",
                        current,
                        max: count
                    });
                }
                current++;

                const { keys, values, tname, mname, deck, qfmt, tags } = stmt.getAsObject();
                const vs = (values as string).split("\x1f");

                const data = {} as any;
                (keys as string).split("\x1f").forEach((k, i) => {
                    data[k] = vs[i];
                });

                let front = simpleMustacheRender(qfmt as string, data);
                if (front === simpleMustacheRender(qfmt as string, {})) {
                    continue;
                }

                front = `@md5\n${md5hasher(this.convertLink(front))}`;

                if (frontSet.has(front)) {
                    continue;
                }
                frontSet.add(front);

                let tag = (tags as string).split(" ");
                tag = tag.filter((t, i) => t && tag.indexOf(t) === i);

                entries.push({
                    deck: (deck as string).replace(/::/g, "/"),
                    model: mname as string,
                    template: tname as string,
                    entry: vs[0],
                    data,
                    front,
                    tag,
                    sourceId
                });
            }
        })();

        (() => {
            const batch = 1000;
            const total = entries.length;
            let subList = entries.splice(0, batch);
            let from = 0;

            while (subList.length > 0) {
                this.callback({
                    text: "Uploading notes",
                    current: from,
                    max: total
                });

                dst.insertMany(subList);
                subList = entries.splice(0, batch);
                from += batch;
            }
        })();
    }

    public close() {
        fs.unlinkSync(this.filepath);
        this.callback({});
        this.db.close();
    }

    private convertLink(s: string) {
        return s.replace(/(?:(?:href|src)=")([^"]+)(?:")/, (m, p1) => {
            return `/media/${this.mediaNameToId[p1]}`;
        });
    }
}

export function md5hasher(s: string | Buffer) {
    return crypto.createHash("md5").update(s).digest("hex");
}